lib/src/parsers/phone_parser.dart

```diff
@@ -168,14 +128,45 @@ class PhoneParser {
   ///
   /// throws a PhoneNumberException if the country calling code is invalid
   @internal
-  static PhoneNumber fromRaw(String phoneNumber) {
+  static PhoneNumber fromRaw(String phoneNumber, { IsoCode? callerCountry, IsoCode? destinationCountry }) {
     phoneNumber = TextParser.normalize(phoneNumber);
-    phoneNumber = InternationalPrefixParser.removeInternationalPrefix(
-      phoneNumber,
-    );
-    final countryCode = CountryCodeParser.extractCountryCode(phoneNumber);
+    String withoutIntlPrefix;
+    if (callerCountry != null) {
+      // as we know the caller country, we can remove the prefix more exactly
+      var metadata = MetadataFinder.getMetadataForIsoCode(callerCountry);
+      withoutIntlPrefix =
+          InternationalPrefixParser.removeInternationalPrefix(
+              phoneNumber, metadata: metadata);
+      if (withoutIntlPrefix.length == phoneNumber.length) {
+        // the given number was not an international number,
+        // but since we know the caller country we may reassemble this for further processing
+        withoutIntlPrefix = metadata.countryCode + NationalNumberParser.removeNationalPrefix(withoutIntlPrefix, metadata);

```

Comment:
If you have `PhoneNumber.fromRaw('+33 655 5705 76', callerCountry: IsoCode.DE)`

then we will reach here with `withoutIntlPrefix = '33'  +  '33 655 5705 76'`, since the country code was not removed.

I'm lacking sleep right now so I'll have to repass on this later, it looks to me that this is missing:

```
    final countryCode = CountryCodeParser.extractCountryCode(phoneNumber);
```

If that was intended by you, this makes the code hard to understand.  It's not `withoutIntlPrefix`, it is `withoutIntlPrefixAndMaybeWithCountryCodeOrDoubleCountryCode` which is not reasonable.
I'm lacking sleep but I have an hard time understanding the rest given this line because there is too much implications / lack of clear intent.

I don't think I'm being pedantic here, I'd like to reduce the cognitive load required here.

Reply:
Not really, if you enter +49... you won't reach this, this is part is only reached for national phone numbers (no exit code was removed).
I adapted the comments to make this more clear.

Reply:
Just so we are on the same page:

It is:  00 +33 475 => international prefix: 00, countryCode: 33 nsn: 475
Or it is:  0 475 => national prefix: 0, nsn: 475.

If   '49 655 5705 76' is provided, wouldn't it reach:

`metadata.countryCode + NationalNumberParser.removeNationalPrefix(withoutIntlPrefix, metadata)`

==

`49 + 49 655 5705 76`

Reply:
Sure, if you provide 49655560576 and specify that this is a German number then it will evaluate this to +49496555705 (which is a legitimate German number), which is just what I would expect.
If you provide 0049655560576 or +4965560576 the routine will realize that an internation prefix was removed and the provided number is an international number.

I agree that the expected format in fromRaw is not completely consistent through this change any more. Previously any existing intl. prefix was just removed blindly, so it made no difference if you provide 0049123, 49123 or +49123. But then you cannot tell local and international numbers apart.

Reply:
> +49496555705 (which is a legitimate German number), which is just what I would expect.

Maybe we are misunderstanding each other. If that happens for german numbers it will happen for other countries too: 

For this test for example:

```
        expect(PhoneNumber.fromRaw('+33 655 5705 76', callerCountry: IsoCode.DE).international,
 equals('+33655570576'));
```

At this line the phone number will be `49 33 655 5705 76`, which I find very odd. The test passes because of the code following this line but this line is confusing for that reason. I'm afraid this might cause issues which were not caught by the tests.

Reply:
In this test that code won't be reached. In that test a +33 ... is given, so it's detected to be international. If the test would just pass in 33... then the scenario you pictured would happen.

Reply:
(Addon hint: Your InternationalPrefixParser.removeInternationalPrefix removes both cases, leading with + and leading country-specific international exit code, so in both cases the national number handling that I added won't be used)

Reply:
Okay, so I had some sleep :D 

This makes much more sens now, I think I just needed to sleep a bit

Reply:
Sleeping helps a lot, same here, so no worries :)
I had similar thoughts about what to do if a national prefix is absent. My problem here is pretty much the same, I know how it works in Germany, but have only little knowledge if that works the same or at least similarly in other countries. From a German perspective: If there is no national prefix then the number is locally dispatched within the "area" where you are located (your mobile network or your city), since that is absent when parsing such a number in fromRaw it's impossible to rebuild the full number with area code then. Police, operator calls etc. might be exceptions to that. From that point of view I would therefore tend to raise an exception in such a case.

Reply:
I believe it would be better to throw an error for the time being. An incomplete behavior is better than a faulty one so people don't rely on it.

Reply:
thank you